go/vt/schemadiff/table.go

```diff
@@ -1319,12 +1322,22 @@ func (c *CreateTableEntity) diffConstraints(alterTable *sqlparser.AlterTable,
 	// evaluate dropped constraints
 	//
 	for _, t1Constraint := range t1Constraints {
-		if _, ok := t2ConstraintsMap[normalizeConstraintName(t1Name, t1Constraint)]; !ok {
+		// Due to how we normalize the constraint string (e.g. in ConstraintNamesIgnoreAll we
+		// completely discard the constraint name), it's possible to have multiple constraints under
+		// the same string. Effectively, this means the schema design has duplicate/redundant constraints,
+		// which of course is poor design -- but still valid.
+		// To deal with dropping constraints, we need to not only account for the _existence_ of a constraint,
+		// but also to _how many times_ it appears.
+		constraintName := normalizeConstraintName(t1Name, t1Constraint)
+		if t2ConstraintsCountMap[constraintName] == 0 {
 			// constraint exists in t1 but not in t2, hence it is dropped
 			dropConstraint := dropConstraintStatement(t1Constraint)
 			alterTable.AlterOptions = append(alterTable.AlterOptions, dropConstraint)
+		} else {
+			t2ConstraintsCountMap[constraintName]--
 		}
 	}
+	// t2ConstraintsCountMap should not be used henceforth.
```

Comment:
Should we make it nil here then as well so that it cannot be used?

Reply:
Good idea. So sorry to have merged this before noticing this comment. I'll address this elsewhere. I just happen to have https://github.com/vitessio/vitess/pull/14373 as an open PR which I'll piggyride.

Reply:
Haha. The linter doesn't like it. If I add `t2ConstraintsCountMap = nil` I get:

```
go/vt/schemadiff/table.go:1341:2: ineffectual assignment to t2ConstraintsCountMap (ineffassign)
	t2ConstraintsCountMap = nil
```

Because nothing _does_ use the map. The linter is sometimes overzealous.

Reply:
@mattlord how do you feel about adding a new cope block and a copy of the map, like so:

```go
	{
		t2ConstraintsCountMapCopy := maps.Clone(t2ConstraintsCountMap)
		for _, t1Constraint := range t1Constraints {
			// Due to how we normalize the constraint string (e.g. in ConstraintNamesIgnoreAll we
			// completely discard the constraint name), it's possible to have multiple constraints under
			// the same string. Effectively, this means the schema design has duplicate/redundant constraints,
			// which of course is poor design -- but still valid.
			// To deal with dropping constraints, we need to not only account for the _existence_ of a constraint,
			// but also to _how many times_ it appears.
			constraintName := normalizeConstraintName(t1Name, t1Constraint)
			if t2ConstraintsCountMapCopy[constraintName] == 0 {
				// constraint exists in t1 but not in t2, hence it is dropped
				dropConstraint := dropConstraintStatement(t1Constraint)
				alterTable.AlterOptions = append(alterTable.AlterOptions, dropConstraint)
			} else {
				t2ConstraintsCountMapCopy[constraintName]--
			}
		}
	}
```
?